File: rust/examples/linux/power/ina226/demo/src/main.rs
use linux_embedded_hal::I2cdev;
use periph::chips::power::{Ina226Full, BOL, BUL};
use std::thread::sleep;
use std::time::Duration;

fn main() {
    let i2c_bus: u8 = std::env::var("I2C_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(1);
    let addr: u8 = std::env::var("I2C_ADDR")
        .ok()
        .and_then(|v| u8::from_str_radix(v.trim_start_matches("0x"), 16).ok())
        .unwrap_or(0x40);

    let dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus");
    let mut chip = Ina226Full::new(dev, addr, 0.1, 2.0).expect("init INA226");

    // Average 16 samples to smooth out noise on a noisy power rail
    chip.configure(3, 4, 4, 7).unwrap();

    // Alert if bus voltage goes outside 4.5–5.5 V (expected 5 V USB supply)
    chip.set_alert(BOL, 5.5, false, false).unwrap();
    chip.set_alert(BUL, 4.5, false, false).unwrap();

    // Sample every second; a real application would sleep between conversions
    for _ in 0..10 {
        while !chip.conversion_ready().unwrap() {
            sleep(Duration::from_millis(10));
        }

        let v = chip.voltage().unwrap();
        let i = chip.current().unwrap();
        let p = chip.power().unwrap();
        println!("V={:.3}V  I={:.4}A  P={:.4}W", v, i, p);

        if chip.overflow().unwrap() {
            eprintln!("WARNING: math overflow — increase max_current or reduce shunt");
        }

        sleep(Duration::from_secs(1));
    }
}

File: cpp/examples/linux/power/INA226/demo/main.cpp
#include <cstdio>
#include <cstdlib>
#include <unistd.h>
#include "I2CConnectionLinux.h"
#include "INA226.h"

int main() {
    const char* bus_env  = getenv("I2C_BUS");
    const char* addr_env = getenv("I2C_ADDR");
    int     bus  = bus_env  ? atoi(bus_env)       : 1;
    uint8_t addr = addr_env ? (uint8_t)strtol(addr_env, nullptr, 0) : 0x40;
    I2CConnectionLinux connection(bus, addr);

    INA226Full ina(connection);                                             // Create INA226 driver, (connection)

    // --- Power rail monitor with over-current alert ---
    // Samples every 500 ms, prints a warning when current exceeds 1 A.
    ina.configure(3, 4, 4, 7);                                            // Configure ADC, (avg=8, 1.1ms, 1.1ms, continuous) → void
    while (true) {
        float v = ina.voltage(), i = ina.current(), p = ina.power();      // Read bus voltage/current/power, () → float
        printf("%.4f V  %.4f A  %.4f W%s\n",
               v, i, p, i > 1.0f ? "  [OC]" : "");
        usleep(500000);
    }
    return 0;
}
